### PR TITLE
Output a better error message when the initial disc is invalid

### PR DESCRIPTION
**Pull request by mguaypaq**
## Description
When calling `vertebral_detection` with an invalid initial disc (that is, a disc label that's not in the PAM50 template), the result is a very unfriendly error message:

```python
Traceback (most recent call last):
  File "/icometrix-install/spinalcordtoolbox-5.3.0/spinalcordtoolbox/vertebrae/core.py", line 206, in vertebral_detection
    approx_distance_to_next_disc = list_distance[list_disc_value_template.index(current_disc - 1)]
ValueError: -2 is not in list
During handling of the above exception, another exception occurred:
Traceback (most recent call last):
  File "/icometrix-install/spinalcordtoolbox-5.3.0/spinalcordtoolbox/scripts/sct_label_vertebrae.py", line 482, in <module>
    main(sys.argv[1:])
  File "/icometrix-install/spinalcordtoolbox-5.3.0/spinalcordtoolbox/scripts/sct_label_vertebrae.py", line 419, in main
    verbose=verbose, path_template=path_template, path_output=path_output, scale_dist=scale_dist)
  File "/icometrix-install/spinalcordtoolbox-5.3.0/spinalcordtoolbox/vertebrae/core.py", line 208, in vertebral_detection
    logger.warning('Disc value not included in template. Using previously-calculated distance: %s', approx_distance_to_next_disc)
UnboundLocalError: local variable 'approx_distance_to_next_disc' referenced before assignment
```

(This can only happen in the first iteration through the loop in `vertebral_detection()`. In further iterations, the error is caught, a warning is displayed, and the previous iteration's value of `approx_distance_to_next_disc` is used.)

This PR adds a check to detect this condition up-front and output a (hopefully) friendlier message to the user.

This is not a complete fix for the linked issue (#3673, see also [the original forum post](https://forum.spinalcordmri.org/t/inconsistency-in-spinalcord-segmentation-leads-to-error-in-in-sct-label-vertebrae-version-5-3-0/809)), but is likely the best we can do in time for the 5.6 release.

## Linked issues
Part of #3673.

### CODE REVIEW

**Comment by mguaypaq** [comment]
Looks like `test_sct_label_vertebrae_high_value_warning` is failing. I'll investigate.

**Comment by mguaypaq** [comment]
Ok, this is ready for review again. I'm very open to the suggestion that we shouldn't merge the new test cases into master, but I wanted to include them in the PR mostly as an illustration of various corner cases (and proof that the new errors/warnings do get triggered).

I tried to plug as many holes as possible with decent error messages, but in my mind this further cements the idea that we should take a good look at the vertebral detection code after the 5.6 release.

File: testing/cli/test_cli_sct_label_vertebrae.py
Comment (joshuacwnewton) [inline testing/cli/test_cli_sct_label_vertebrae.py:54]: I really love the in-depth descriptions you put into your commits. For example, the description in https://github.com/spinalcordtoolbox/spinalcordtoolbox/commit/a442e1291c060600bed126aeef617737c0296cd6 really aided me when trying to trace the behavior of `sct_label_vertebrae`.

I'm wondering... would you be able to adapt these descriptions into docstrings for the tests? That way, we have a loose description of the script's behavior for each corner case test (i.e. -1, 0, 19, 20, 21) attached to the tests themselves.

**Review by joshuacwnewton — APPROVED** [review]
> I'm very open to the suggestion that we shouldn't merge the new test cases into master, but I wanted to include them in the PR mostly as an illustration of various corner cases (and proof that the new errors/warnings do get triggered).

I found these corner case tests to be _very_ valuable for reviewing. 

(What I did locally was to rebase the "adding tests" commit to be the first commit, then step through your changes to see what the effect on the tests would be before/after, making it act as a sort of TDD-esque changeset.) 

* test_sct_label_vertebrae_initial_disc_no_inferior: ValueError: 20 is not in list (Fixed by a442e12) 
* test_sct_label_vertebrae_initial_disc_zero: spinalcordtoolbox.vertebrae.core.EmptyArrayError: Center of mass can't be calculated on empty arrays. (Fixed by 16dbd6f)
* test_sct_label_vertebrae_initial_disc_too_low: ValueError: min() arg is an empty sequence (Fixed by b8e146f) 
* test_sct_label_vertebrae_initial_disc_too_high: ValueError: 21 is not in list (Fixed by b8e146f)

So, thank you for including these tests. :) 

I've included one comment, but it's just about documentation for the tests, so I'm going to pre-approve.

File: testing/cli/test_cli_sct_label_vertebrae.py
Comment (joshuacwnewton) [inline testing/cli/test_cli_sct_label_vertebrae.py:73]: Actually, wait, sorry. It occurs to me that none of the error messages observed from the tests (https://github.com/spinalcordtoolbox/spinalcordtoolbox/pull/3775#pullrequestreview-950071054) match the error message in https://github.com/spinalcordtoolbox/spinalcordtoolbox/pull/3775#issue-1210242621.

Did we want to try and create a test that (would have) thrown the `UnboundLocalError`?

File: spinalcordtoolbox/vertebrae/core.py
Comment (joshuacwnewton) [inline spinalcordtoolbox/vertebrae/core.py:211]: Ah! So, there was a small, curious thing that caught my attention:

* The error I observed from [`test_sct_label_vertebrae_initial_disc_too_low`](https://github.com/spinalcordtoolbox/spinalcordtoolbox/pull/3775#pullrequestreview-950071054) prior to your try/except changes was `ValueError: min() arg is an empty sequence`
* However, the user's error from [the reported traceback](https://github.com/spinalcordtoolbox/spinalcordtoolbox/pull/3775#issue-1210242621) was an `UnboundLocalError`

Either way, your fix _does_ address the issue (by catching a `current_value` that doesn't exist in the template). But, the test itself produced slightly different behavior _prior_ to your changes. Curious!

I tried investigating further by tracing your test and seeing how it differs from the user's report. And it turns out, when we supply `-initz 40,-1` as we do in that test, the `current_value` actually gets parsed as _254_???

image

I think this is because of this line:

https://github.com/spinalcordtoolbox/spinalcordtoolbox/blob/bad57807faf05643e2697c911f3403c4f57d9f66/spinalcordtoolbox/scripts/sct_label_vertebrae.py#L383

Which creates a `uint8` label image from the `initz` argument (i.e. the image is bound to `[0, 255]`). So, I'm guessing that by supplying -1, it wraps around and creates a labeled image with the value `254`? (I tried changing the test to -2 and it produced an image with label value `253`.)

But, as I mentioned, your check still catches the value `254` as it would `-1`, so... I'm not sure. Is this something we want to investigate while we're here? If not, this has my approval. :)

Comment (joshuacwnewton) [reply spinalcordtoolbox/vertebrae/core.py:211]: Oh wait, I can just quickly update `create_labels_along_segmentation` to not use `uint8`, to see what would happen if we did in fact supply `-1`. One sec. :V

Comment (joshuacwnewton) [reply spinalcordtoolbox/vertebrae/core.py:211]: So, if I change this line to use `dtype=np._int`:

https://github.com/spinalcordtoolbox/spinalcordtoolbox/blob/bad57807faf05643e2697c911f3403c4f57d9f66/spinalcordtoolbox/labels.py#L111

We get the `-1` label in the image as expected. However, then when the image gets dilated here:

https://github.com/spinalcordtoolbox/spinalcordtoolbox/blob/bad57807faf05643e2697c911f3403c4f57d9f66/spinalcordtoolbox/scripts/sct_label_vertebrae.py#L402

The dilation causes the `-1` value to disappear. So, again, another little factor preventing your `-1` test from actually doing what it's supposed to. :P

Comment (joshuacwnewton) [reply spinalcordtoolbox/vertebrae/core.py:211]: Ah, we use [`skimage.morphology.dilation`](https://scikit-image.org/docs/dev/api/skimage.morphology.html#skimage.morphology.dilation) for our dilate function, which inherently returns a `uint8` image.

So... I guess negative label values are impossible via `fname_labelz`? But... the user's error certainly suggests that it's possible...

Comment (joshuacwnewton) [reply spinalcordtoolbox/vertebrae/core.py:211]: Again though, just to re-emphasize, I think we've covered all of our bases with the new corner cases (0, 19, 20, 21), so fixating on the `-1` test isn't really that important?

I just wanted to highlight that the test wasn't doing what we thought it was doing. :P

Comment (mguaypaq) [reply spinalcordtoolbox/vertebrae/core.py:211]: That's a great investigation! I didn't think about numpy's wrap-around integer types here, but it's a good source of surprises and bugs. And you're right that image dilation does surprising things with negative numbers.

But! There's one more place where negative numbers can creep in :stuck_out_tongue: After straightening of `fname_labelz`, the label is read from here:
https://github.com/spinalcordtoolbox/spinalcordtoolbox/blob/bad57807faf05643e2697c911f3403c4f57d9f66/spinalcordtoolbox/scripts/sct_label_vertebrae.py#L414
and then we subtract 1 before handing to `vertebral_detection`:
https://github.com/spinalcordtoolbox/spinalcordtoolbox/blob/bad57807faf05643e2697c911f3403c4f57d9f66/spinalcordtoolbox/scripts/sct_label_vertebrae.py#L433
Note that by this point `init_disc[1]` is an honest-to-goodness Python int, because it's explicitly cast here:
https://github.com/spinalcordtoolbox/spinalcordtoolbox/blob/bad57807faf05643e2697c911f3403c4f57d9f66/spinalcordtoolbox/vertebrae/core.py#L298

So, I think `-1` is in fact the *only* possible negative value, and it happens with `-initz *,0` (or I guess, any positive or negative multiple of 256?). This also explains why you saw 254 instead of 255 when using `-initz *,-1`.

----

~~Now, I want to argue that the `min() arg is an empty sequence` is actually already addressed by this PR.~~ The error is raised on this line after breaking out of the main loop:
https://github.com/spinalcordtoolbox/spinalcordtoolbox/blob/3f7832ad5670fc6cb875c8b413a37c15a1e5166c/spinalcordtoolbox/vertebrae/core.py#L232
The argument to `min` can only be empty if this code in the loop never executes:
https://github.com/spinalcordtoolbox/spinalcordtoolbox/blob/3f7832ad5670fc6cb875c8b413a37c15a1e5166c/spinalcordtoolbox/vertebrae/core.py#L168-L176
And the only `break` before that is here:
https://github.com/spinalcordtoolbox/spinalcordtoolbox/blob/3f7832ad5670fc6cb875c8b413a37c15a1e5166c/spinalcordtoolbox/vertebrae/core.py#L145-L151
However, ~~the new error checking code prevents this from happening on the first iteration,~~ because we now check whether the initial `current_value` is there:
https://github.com/spinalcordtoolbox/spinalcordtoolbox/blob/3f7832ad5670fc6cb875c8b413a37c15a1e5166c/spinalcordtoolbox/vertebrae/core.py#L134-L136
Edit: the check which would actually prevent this is `current_value in range(-1-len(list_disc_value_template), len(list_disc_value_template))`. But for the PAM50 template, `list_disc_value_template == [0, 1, 2, ..., 19]`, so the new error checking code accidentally catches this condition.

I think we're hitting diminishing returns here; the real solution is not to add more error checking code, but to actually fix the underlying code after the 5.6 release. So, I think I'll add some comments and/or docstrings for the new corner-case tests and call this PR done, if that's ok with you.

Comment (joshuacwnewton) [reply spinalcordtoolbox/vertebrae/core.py:211]: >  So, I think I'll add some comments and/or docstrings for the new corner-case tests and call this PR done, if that's ok with you.

Perfectly reasonable to me! That was very much my feelings too after writing https://github.com/spinalcordtoolbox/spinalcordtoolbox/pull/3775#discussion_r856302643.

**Review by joshuacwnewton — APPROVED** [review]

**Pull request merged**
